src/main/java/com/segment/analytics/android/integrations/optimizelyx/OptimizelyXIntegration.java

```diff
@@ -1,63 +1,170 @@
+  private List<TrackPayload> trackEvents = new ArrayList<>();
+
+  public static Factory factory(OptimizelyManager manager) {
+    return new Factory(manager);
+  }
+
+  private static class Factory implements Integration.Factory {
+    private final OptimizelyManager manager;
+
+    Factory(OptimizelyManager manager) {
+      this.manager = manager;
+    }
+
+    @Override
+    public Integration<?> create(ValueMap settings, Analytics analytics) {
+      Logger logger = analytics.logger(OPTIMIZELYX_KEY);
+      return new OptimizelyXIntegration(analytics, manager, settings, logger);
+    }
+
+    @Override
+    public String key() {
+      return OPTIMIZELYX_KEY;
+    }
+  }
+
+  public OptimizelyXIntegration(
+          final Analytics analytics, OptimizelyManager manager, ValueMap settings, Logger logger) {
+    this.client = manager.getOptimizely();
     this.logger = logger;
+
+    trackKnownUsers = settings.getBoolean("trackKnownUsers", false);
+    listener = new OptimizelyNotificationListener(analytics);
+
+    if (client.isValid()) {
+      isClientValid = true;
+      logger.verbose("OptimizelyClient instance is valid. Ensure you have initialized Optimizely.");
+      client.addNotificationListener(listener);
+    } else {
+      logger.verbose("OptimizelyClient instance is invalid.");
+      pollOptimizelyClient();
+    }
+  }
+
+  @Override
+  public void track(TrackPayload track) {
+    super.track(track);
+
+    if (isClientValid) {
+      String id;
+      String userId = track.userId();
+      String event = track.event();
+      Map<String, String> properties = track.properties().toStringMap();
+
+      if (trackKnownUsers && isNullOrEmpty(userId)) {
+        logger.verbose(
+                "Segment will only track users associated with a userId "
+                        + "when the trackKnownUsers setting is enabled.");
+        return;
+      } else if (trackKnownUsers && !isNullOrEmpty(userId)) {
+        id = track.userId();
+      } else {
+        id = track.anonymousId();
+      }
+      client.track(event, id, properties);
+      logger.verbose("client.track(%s, %s, %s)", event, id, properties);
+    } else {
+      trackEvents.add(track);
+      logger.verbose("This event was pushed in an array " + track);
+      for (TrackPayload t : trackEvents) {
+        logger.verbose(t.toString());
+      }
+    }
+  }
+
+  @Override
+  public void reset() {
+    super.reset();
+
+    client.removeNotificationListener(listener);
+    logger.verbose("client.removeNotificationListener(listener)");
   }
 
-  @Override public void identify(IdentifyPayload identify) {
+  private void pollOptimizelyClient() {
+    final Handler handler = new Handler();
+    handler.postDelayed(new Runnable() {
+      @Override
+      public void run() {
+        if (!client.isValid()) {
+          logger.verbose("The OptimizelyClient instance is invalid. Please check your Optimizely project id.");
+          try {
+            Thread.sleep(60000);
```

Comment:
Why pause execution for this time interval?

Reply:
Recursive function that polls the validity of the optimizelyClient once per minute until it's valid. This value is arbitrary, but we should seek feedback from Jaime on a more appropriate interval.